Mathlib/CategoryTheory/Shift/Opposite.lean

```diff
@@ -173,4 +180,32 @@ noncomputable def commShiftUnop
 
 end Functor
 
+namespace Adjunction
```

Comment:
There is a useful general statement about natural transformation, and the case of adjunctions follows from it. The structure of the code should be as follows (the definition of the opposite of an adjunction should be fixed...):
```lean
namespace NatTrans

attribute [local instance] Functor.commShiftOp

variable {F} {G : C ⥤ D} [F.CommShift A] [G.CommShift A]

lemma commShift_op (τ : F ⟶ G) [NatTrans.CommShift τ A] :
    NatTrans.CommShift (C := OppositeShift C A) (D := OppositeShift D A) (NatTrans.op τ) A := by
  sorry

end NatTrans

namespace Adjunction

attribute [local instance] Functor.commShiftOp NatTrans.commShift_op

variable {F} {G : D ⥤ C} (adj : F ⊣ G)

-- to be moved to `CategoryTheory.Adjunction.Opposites` as a replacement
-- of the broken `opAdjointOpOfAdjoint` (Note: this definition made
-- sense before #16317 when adjunctions where defined in mathlib with data
-- including the unit, the counit and the `homEquiv`, but now 
-- that the `homEquiv` field has been removed, the following would be
-- much better)
@[simps]
def op : G.op ⊣ F.op where
  unit := NatTrans.op adj.counit
  counit := NatTrans.op adj.unit
  left_triangle_components _ := Quiver.Hom.unop_inj (by simp)
  right_triangle_components _ := Quiver.Hom.unop_inj (by simp)

lemma commShift_op [F.CommShift A] [G.CommShift A]  [adj.CommShift A] :
    Adjunction.CommShift (C := OppositeShift D A) (D := OppositeShift C A)
      adj.op A where
  commShift_unit := by dsimp; infer_instance
  commShift_counit := by dsimp; infer_instance

end Adjunction
```

Reply:
Okay, two questions:
- First, I tried of course to do something similar with pulled back shifts, but I run into annoying problems. I could fix them by adding an `@[ext]` attribute to `Functor.CommShift` and using it to prove lemmas like `Functor.CommShift.id (PullbackShift C φ) = Functor.commShiftPullback φ (𝟭 C)` (and similarly for compositions), however, I am wondering whether there is a simpler way.
- Second, if I understand well you want me to edit the file `CategoryTheory.Adjunction.Opposites` and change the definitions of `opAdjointOpOfAdjoint` (and, I presume, also `adjointOfOpAdjointOp`). However, do you also want to change the names? There are many options for "opposite adjunctions" with a complicated naming scheme (like `adjointUnopOfAdjointOp`, `opAdjointOfUnopAdjoint` etc), so if you change one name wouldn't you need to change all their names?

Reply:
For the second question, we have `Equivalence.op`, so that `Adjunction.op` seems a good name to me, and similarly `Adjunction.unop`. I think we can get rid of the variants `adjointUnopOfAdjointOp` and `opAdjointOfUnopAdjoint` which are basically unused (you may introduce deprecated aliases referring the more relevant new `op` or `unop`). (We probably have these variants because `F.op.unop` was not defeq to `F` in Lean 3, but it is now.) It may be better to do that in a separate PR.

For the first question, I do not have time to look at the details now, but it would be better if we can avoid proving equalities between `Functor.CommShift`s. I am not sure it is relevant, but it may be possible to introduce an isomorphism between two copies of the identity (with different `CommShift`s on it) and show a `NatTrans.CommShift` for it?

Reply:
The problem seems to be that Lean does not think that the identity of `C` and the identity of `PullbackShift C φ` are the same, so this solution might work, yes.